FIREFLY-1358: Convert Range Slider and Version Info to JoyUI

kpuriIpac opened the pull request:
#### [Firefly-1358](https://jira.ipac.caltech.edu/browse/FIREFLY-1358)

**Updates 12/12**
- converted as much of `FitsRotationDialog` and `ColorTableDropDownView` to joy UI as possible. Tried to remain consistent in the changes (and the overall look of the UI as well) 
- The Joy UI slider accepts a different format of `marks` object (array or bool) and I see that previously we had a check for it in the `checkMarksObject` function in `RangeSliderView.jsx`. Is this still necessary? 
- VersionInfo is pretty much the same as before. Attempted to spend some time converting `VersionInfo.jsx` to use Joy's `Grid` component, but I struggled a bit with getting this right. 
  - If it's a good idea to use `Grid`, I can spend some more time tomorrow to get it right/working. Other option is to come back to this part of `VersionInfo.jsx` when we start using Grid more widely so I have a better idea of how to use it right. 

**Testing**: 
- https://fireflydev.ipac.caltech.edu/firefly-1358-range-slider-joyui/firefly
- `VersionInfo` testing: 
  - Click on the firefly logo to get the Firefly Version Info dialog, make sure it looks ok. 
- `FitsRotationDialog` testing: 
  - Search for an image, click on the fits rotation dialog from our tools, ensure it looks okay. Use the slider and the input box to update the values. 
- `ColorTableDropDownView` testing:
  - Search for a fits image, say m81. Test the color table dropdown, and make sure it looks okay. You should clearly see `Color Bar`, `Bias` and `Contrast`. 
  - Search for another image, but this time for image type, choose `Create 3-color Composite`. Search three different times selecting red, green and blue. 
    - Click on the color dropdown now, you should see `RED Bias, RED Contrast `, `GREEN Bias, GREEN Contrast`, `BLUE Bias, BLUE Contrast` clearly. Play around with the slider. 
  - **_Important_**: For both of these dropdowns, click on the little popout arrow (top right) to open the dropdown as a PopupPanel and ensure the UI still looks okay.

loitly commented on src/firefly/js/ui/VersionInfo.jsx:
```diff
@@ -52,14 +55,20 @@ function VersionInfoFull() {
-                <div style={{display: 'flex', justifyContent: 'space-evenly', marginLeft:60, alignItems:'flex-start'}}>
+        <Stack justifyContent='space-evenly' direction='column' sx={{fontSize: 14, width: '100%'} }>
+            <Stack direction='row'>
+                <Stack direction='column' alignItems='center' sx={{pr:0.6, mt:-2} }>
```
- `pr:0.6` is not needed.
- `mt:-2` is to fix the `marginTop:10` set in its child img.  You can remove both instead.

loitly commented on src/firefly/js/ui/VersionInfo.jsx:
```diff
@@ -68,10 +77,10 @@ function VersionInfoFull() {
-                                     value={versionAsText} size={16} buttonStyle={{backgroundColor: 'unset'}}/>
-                </div>
-            </div>
+                    <Typography endDecorator={<CopyToClipboard style={{justifySelf: 'start', marginLeft:10}}
```
Why is `Typography` used?

loitly reviewed (commented):
I am nitpicking.  It was useful for me as well to familiarize myself with the library.
In VersionInfo, a Grid is probably better suited then nested Stacks.
In general, I think you should always try to keep it simple when complex is not needed. (addition by subtraction)

kpuriIpac commented:
> I am nitpicking. It was useful for me as well to familiarize myself with the library. In VersionInfo, a Grid is probably better suited then nested Stacks. In general, I think you should always try to keep it simple when complex is not needed. (addition by subtraction)

Yes please, I think this is very helpful for me too at least in the initial Joy UI PRs so that there's also more consistency across our changes!

kpuriIpac commented on src/firefly/js/ui/VersionInfo.jsx:
```diff
@@ -82,26 +91,27 @@ function VersionInfoFull() {
 const Acknowledgement= () => (
-    <div style={{padding:'10px 5px 3px 5px', width:520, fontSize:'85%', lineHeight:1.3}}>
+
+    <Typography level='body-xs' width={'100%'} sx={{pt:0.6,pr:0.3, pb:0.2, pl:0.3, lineHeight:1.3}}>
```
I was trying to replicate the original `padding:'10px 5px 3px 5px'` but yes i see it's just overkill. I'll switch to `pt:1 `instead.

robyww commented on src/firefly/js/ui/VersionInfo.jsx:
```diff
@@ -82,26 +91,27 @@ function VersionInfoFull() {
-        <a href='https://www.caltech.edu' target='caltech-window'>Caltech</a>
+        <Link underline='always' color='primary' href='https://www.caltech.edu' target='caltech-window'>Caltech</Link>
             &nbsp;has been supported by NASA, principally through&nbsp;
-        <a href='https://irsa.ipac.caltech.edu' target='ipac-window'>IRSA</a>
+        <Link underline='always' color='primary' href='https://irsa.ipac.caltech.edu' target='ipac-window'>IRSA</Link>
            , and by the National Science Foundation, through the&nbsp;
-        <a href='https://www.lsst.org/' target='rubin-window'>Vera C. Rubin Observatory</a>
+        <Link underline='always' color='primary' href='https://www.lsst.org/' target='rubin-window'>Vera C. Rubin Observatory</Link>
             . Firefly is open-source software, available on&nbsp;
-        <a href='https://github.com/Caltech-IPAC/firefly' target='github-window'>GitHub</a>
+        <Link underline='always' color='primary' href='https://github.com/Caltech-IPAC/firefly' target='github-window'>GitHub</Link>
         &nbsp;and&nbsp;
-        <a href='https://hub.docker.com/repository/docker/ipac/firefly' target='dockerhub-window'>DockerHub</a>
+        <Link underline='always' color='primary' href='https://hub.docker.com/repository/docker/ipac/firefly' target='dockerhub-window'>DockerHub</Link>
```
Modify `ThemeSetup.js` to setup the default link. then take out the properties
from `<Link/>`
```js
 JoyLink: {
     defaultProps: {
        underline: 'always',
        color: 'primary'
  }
```

robyww commented on src/firefly/js/ui/VersionInfo.jsx:
```diff
@@ -52,14 +55,20 @@ function VersionInfoFull() {
-                    <a href='https://www.ipac.caltech.edu' target='ipac-window'><img alt='IPAC' src={IPAC_LOGO}/></a>
-                </div>
-                <div style={{display: 'flex', justifyContent: 'space-evenly', marginLeft:60, alignItems:'flex-start'}}>
+        <Stack justifyContent='space-evenly' direction='column' sx={{fontSize: 14, width: '100%'} }>
```
you can use `level` on `<Typography/>`. We want to use the JoyUI fonts sizes.

robyww commented on src/firefly/js/ui/VersionInfo.jsx:
```diff
@@ -35,10 +38,10 @@ export function getVersionInfoStr(includeBuiltOnDate, includeBuildType) {
 const Entry = ({desc, value}) =>
-    (<React.Fragment>
+    (<>
         <div className='Version-grid-desc'>{desc}</div>
         <div className='Version-grid-value'>{value}</div>
-    </React.Fragment>);
+    </>);
```
These should be `<Typography/>`

robyww commented on src/firefly/js/ui/RangeSliderView.jsx:
```diff
@@ -18,30 +20,28 @@ export function RangeSliderView({min=0, max=100, minStop, maxStop, className, ma
-                        onChange={onSliderChange} />
-            </div>
-        </div>
+        <Box {...{style}} >
```
take `sx` instread of `style`

robyww commented on src/firefly/js/ui/RangeSliderView.jsx:
```diff
@@ -67,7 +67,7 @@ RangeSliderView.propTypes = {
 };
 
-
+//TODO: marks previousLy using rc-slider: marks: PropTypes.objectOf(checkMarksObject), so add check for new marks as well?
```
I don't understand the todo

robyww commented on src/firefly/js/visualize/ui/ColorTableDropDownView.jsx:
```diff
@@ -213,41 +223,40 @@ const AdvancedColorPanel= ({allowPopout}) => {
+                <Typography level='body-xs' style={{padding: '5px 0 2px 0'}}>Color Bar</Typography>
                 <RangeSliderView {...{
-                    style:{paddingTop: 7, width: 200},
+                    style:{paddingBottom:3, marginTop:-12, width: 200},
```
when you start doing sx this would be `sx:{bp:.5,mt:-3}`

robyww commented on src/firefly/js/visualize/ui/ColorTableDropDownView.jsx:
```diff
@@ -261,34 +270,38 @@ const AdvancedColorPanel= ({allowPopout}) => {
                            onClick={() => changeBiasContrastColor(colorTableId,bias,contrast,useRed,useGreen,!useBlue)}/>}
             <DropDownVerticalSeparator useLine={true}/>
-            <div style={{width: 230, height: 126 * plotState.getBands().length}}>
+            <Box style={{width: 230, height: 126 * plotState.getBands().length}} pb={1}>
```
box can take the styles as props, like you did with `pb`

robyww commented on src/firefly/js/visualize/ui/ColorTableDropDownView.jsx:
```diff
@@ -261,34 +270,38 @@ const AdvancedColorPanel= ({allowPopout}) => {
                                         </div>
                                         <RangeSliderView {...{
-                                            style:{paddingTop: 7, width: 200},
+                                            style:{paddingTop: 0, marginTop:-12, width: 200},
```
passing sx would be `sw:{pt:0, mt:-3}`

kpuriIpac replied:
I found that this actually works with `mt:-1.5` (12/8) which means it's based on a root font size of 8px.`mt:-3` would work for a root font-size of 4px. I think once we decide on and set a font-size at root, we may need to make changes to some of these `sx` values later? (unless we decide to go with 8px)

robyww commented on src/firefly/js/visualize/ui/ColorTableDropDownView.jsx:
```diff
@@ -261,34 +270,38 @@ const AdvancedColorPanel= ({allowPopout}) => {
-                                            <span style={{color:b.key}}>{b.key}</span> Bias
+                                <Box key={b.key}>
+                                    <Stack alignItems='center' spacing={0}>
+                                        <div style={{margin: `${idx?'10':'5'}px 0 0 0`, paddingTop:idx?3:0, textAlign:'center', alignSelf:'stretch', borderTop:`1px solid rgba(0,0,0,${idx?'.1':'0'}`}}>
```
I think you can use `Divider` here

robyww commented on src/firefly/js/visualize/ui/ColorTableDropDownView.jsx:
```diff
@@ -261,34 +270,38 @@ const AdvancedColorPanel= ({allowPopout}) => {
-                                        <div style={{padding: '30px 0 0 0'}}><span style={{color:b.key}}>{b.key}</span> Contrast</div>
+                                    </Stack>
+                                    <Stack alignItems='center' spacing={0}>
+                                        <Typography level='body-xs' style={{padding: '10px 0 2px 0'}}>
+                                            <Typography level='body-xs' style={{color:b.key}}>{b.key}</Typography> Contrast
```
Typography is one of those CCS Utility components. you can use styles as props.

robyww commented on src/firefly/js/visualize/ui/ColorTableDropDownView.jsx:
```diff
@@ -261,34 +270,38 @@ const AdvancedColorPanel= ({allowPopout}) => {
+                                        </Typography>
                                         <RangeSliderView {...{
-                                            style:{paddingTop: 7, width: 200},
+                                            style:{paddingTop: 0, marginTop:-12, width: 200},
```
`sx`

robyww reviewed (changes requested):
This PR is almost done.  The slider and version panel look good.

The things to change are small.

- use `sx` instead of `style`
- make RangeSliderView take `sx` instead of `style`
- for CSS utility classes (Box,Typography,Stack) put the styles as props.

robyww commented:
one more thing... You might want to default the dialog to be a little wider to handle the large font. Maybe set the width in `em`

kpuriIpac commented:
Thanks @robyww. Made the changes as requested. 

- Using JoyUI's `Divider` in all 3 places in ColorTableDropDown now instead of our `DropDownVerticalSeparator` separator
- In FITs Rotation Dialog, I removed the padding from the top to get rid of some of the whitespace between the title and the input/slider. 
- Got rid of the `rc-slider` package (and updated `package-json-doc.md` as well)
- Attempted to increase the width of the VersionInfo popup, but I noticed that while `showInfoPopup` is called with `sx`, and `PopupPanel` is also called with this same `sx` prop, the `sx` prop is not received/used in the `PopupPanel` component itself, so I'm not sure what to best way to pass width via sx in this case is? 
  -  should we receive it in `PopupPanel` and pass it on to `PopupHeaderTop` to set `width`? 

Could you take a final look at the new test build to make sure the new UI changes look okay? (since I rebased and got your popup UI changes as well). 
  - the VersionInfo dialog 
  - Fits Rotation Dialog 
  - ColorTableDropDown from both 3color and just a regular image search (notice the `Divider` in both)

kpuriIpac commented on yarn.lock:
```diff
@@ -6096,7 +6087,7 @@ rc-tree@~5.7:
     rc-util "^5.16.1"
     rc-virtual-list "^3.5.1"
 
-rc-util@^5.16.1, rc-util@^5.21.0, rc-util@^5.27.0, rc-util@^5.36.0, rc-util@^5.38.0:
+rc-util@^5.16.1, rc-util@^5.21.0, rc-util@^5.36.0, rc-util@^5.38.0:
```
BTW @robyww, I had manually removed `rc-slider` from `package.json` and run `yarn install` to update `yarn.lock`. 

This also ended up removing `rc-util@^5.27.0` and `classnames@^2.2.5` (both versions may be related to `rc-slider`?)...wanted to make sure this is okay.

kpuriIpac commented:
> * Attempted to increase the width of the VersionInfo popup, but I noticed that while `showInfoPopup` is called with `sx`, and `PopupPanel` is also called with this same `sx` prop, the `sx` prop is not received/used in the `PopupPanel` component itself, so I'm not sure what to best way to pass width via sx in this case is?
>   

Hey @robyww, can you specifically check my latest commit(https://github.com/Caltech-IPAC/firefly/pull/1458/commits/001eb3338697a1c5f3dea9945cf492cf567c8f6d) for a fix for this? (I can undo this commit if this is not the way we want to do this), I tried to fix the width for the Version popup. To do this in as generic a way as possible: 

- I set `maxWidth` from `500px` to `100%` instead of the `Stack` in `makeContent` (in PopupUtil.jsx) (otherwise, with `overflow:hidden`), if let's say the `Version` name is too long (in my case it is: 2023.3-DEV:FIREFLY-1358-Range-Slider-JoyUI_f13a) then it gets cut off and we also lose the `CopyToClipboard` icon. 
  - an alternate option for this would be, leave the `maxWidth` as `500px`, but pass in an optional `overflow` value for this `Stack`. For the Version dialog, if `overflow` is set to `visible`, we can get a horizontal scroll instead. 
- in PopupPanel.jsx, I try to get width `from` `sx` (or set it to empty if it does not exist) and pass it along to `PopupHeaderTop` to set the `Card's` `width`

I did notice that in your last PR for Popup Panels ([Firefly-1366 dialogs](https://github.com/Caltech-IPAC/firefly/pull/1462/files#top)) you removed `sx` being passed to `PopupPanel`. I am guessing you had a reason to do that?

Again, this is just me testing this and I tried to keep it generic - but I can reverse this commit if it's messy, let me know what you think! test build is updated so you can see this Version dialog there.

robyww reviewed (commented):
>I set maxWidth from 500px to 100% instead of the Stack in makeContent (in PopupUtil.jsx)....

You can just set it on the class. This is the idea that @loi has been talking about and shows the power of the `sx` system

robyww commented:
My example go lost... here it is again.
```jsx
export const showFullVersionInfoDialog = (title = '') => showInfoPopup( 
<VersionInfoFull/>, 
`${title} Version Information`, 
{'.FF-Popup-Content': {width:'40em'}});
```

kpuriIpac commented:
> 
> You can just set it on the class. This is the idea that @Loi has been talking about and shows the power of the `sx` system

Okay yeah I remember that now, thanks, that makes so much more sense! made this change (still needed to just pass `sx` on to `PopupHeaderTop` from `PopupPanel` to be able to set `width` of the `Box` there via `sx`)

kpuriIpac commented on src/firefly/js/ui/PopupPanel.jsx:
```diff
@@ -100,16 +100,17 @@ PopupPanel.propTypes= {
+function PopupHeaderTop({modal,zIndex,left,top,visibility,ctxRef,dialogMoveStart,dialogMoveEnd,sx,
                             onMouseEnter,onMouseLeave,dialogMove,children,title,askParentToClose}) {
     return (
-        <Box style={{zIndex, position:'relative'}}>
+        <Box sx={sx} style={{zIndex, position:'relative'}}>
```
@robyww this was bothering me for a while so I decided to look into it. Leaving this comment here for when you look at this on Monday. I think the reason we _need_ this here is because: 

- sx is only applied to MUI/JoyUI components directly 
- so while we're calling PopupPanel via sx here:
  - `<PopupPanel title={title} sx={sx}> {makeContent(content,sx)} </PopupPanel>`
  - I don't think that'll do anything until we apply that `sx` to a JoyUI component 
- so applying it to the `Box` here ensures that the `children` (in this case the `Stack` with `className:FF-Popup-Content` gets our change of `{'.FF-Popup-Content': {width:'40em',maxWidth:'100%'}`
- In fact, I think I can make it cleaner by: 
  - `{'.FF-Popup-Content': {maxWidth:'100%'}, '.ff-PopupPanel':{width:'40em'}}`
  - since this ensures that the parent Card component increases its width, and the child (`DialogContent`) shows all of the content by having its `maxWidth:100%`

robyww replied:
oh, your right! I will approve the PR. Just make the change so we don’t use style and `sx`
```js
<Box sx={{…sx, zIndex, position:'relative'}}>
```

robyww reviewed (approved):
just that one change and merge.
